runtime/src/sorted_storages.rs

```diff
@@ -54,9 +54,25 @@ impl<'a> SortedStorages<'a> {
     // 2. slots and source are the same len
     pub fn new_with_slots<'b>(
         source: impl Iterator<Item = (&'a SnapshotStorage, &'b Slot)> + Clone,
+        min_slot: Option<Slot>,
+        // highest valid slot. Only matters if slots array does not contain a slot >= max_slot_inclusive.
+        // An example is a slot that has accounts in the write cache at slots <= 'max_slot_inclusive' but no storages at those slots.
+        // None => self.range.end = slots.max() + 1
+        // Some(slot) => self.range.end = std::cmp::max(slot, slots.max())
+        max_slot_inclusive: Option<Slot>,
     ) -> Self {
         let mut min = Slot::MAX;
         let mut max = Slot::MIN;
+        // none, either, or both of min/max could be specified
+        if let Some(slot) = min_slot {
+            min = std::cmp::min(slot, min);
+            max = std::cmp::max(slot + 1, max);
+        }
+        if let Some(slot) = max_slot_inclusive {
+            min = std::cmp::min(slot, min);
+            max = std::cmp::max(slot + 1, max);
+        }
+
```

Comment:
if we assert here:

```
if let Some(max_slot) = max_slot_inclusive {
        assert!(max_slot >= min_slot)
}
```

Then written this way, this block is easier for me to reason about all the cases:
```
let (mut min, mut max) = match (min_slot, max_slot_inclusive) {
            (Some(min_slot), Some(max_slot_inclusive)) => (min_slot, max_slot_inclusive),
            (Some(min_slot), None) => (min_slot, Slot::MAX),
            (None, Some(max_slot_inclusive)) => (Slot::MIN, max_slot_inclusive + 1),
            (None, None) => (Slot::MIN, Slot::MAX),
        };
```

Reply:
I'd prefer not to add an assert. Also, it seems possible that calculations of min could end up > max given that we have several sources we're merging: global roots, ancestors, current slot of bank?, and the global write cache. We are really looking for the max range that we have to iterate over to check stores or write cache for accounts that we should include in our calculation. 

I did rework the function to make it much more clear what is going on. min_slot and max_slot_inclusive are just 2 more 'slot' values run through the min/max calculation, along with all the slots in the iter.